File: src/patch.rs
use winsafe::{prelude::*, self as w, co};

pub fn is_vscode_running() -> w::SysResult<bool> {
	let hpl = w::HPROCESSLIST::CreateToolhelp32Snapshot(co::TH32CS::SNAPPROCESS, None)?;
	let mut pe = w::PROCESSENTRY32::default();
	let mut found = false;

	for pe in hpl.iter(&mut pe) {
		if pe?.szExeFile() == "Code.exe" {
			found = true;
			break;
		}
	}

	hpl.CloseHandle()?;
	Ok(found)
}

pub fn patch_font(install_dir: &str) -> w::AnyResult<()> {
	let css_path = _build_css_path(install_dir);
	let orig_contents = _read_css_contents(&css_path)?;

	const END_OF_COMMS: &str = "-*/";
	const MAGIC_PATCH: &str = "\n*{text-shadow:transparent 0px 0px 0px, rgba(0, 0, 0, 0.5) 0px 0px 0px !important;}";

	// Find index past the comments block.
	let mut idx_start_code = match orig_contents.find(END_OF_COMMS) {
		Some(idx) => idx,
		None => return Err("CSS end of comments not found.".into()),
	};

	idx_start_code += END_OF_COMMS.len();

	// Is our magic path the first thing past the comments block?
	if MAGIC_PATCH == &orig_contents[idx_start_code..(idx_start_code + MAGIC_PATCH.len())] {
		return Err("Font already patched.\n\nNothing to do.".into());
	}

	let mut new_contents = String::with_capacity(orig_contents.len() + MAGIC_PATCH.len());
	new_contents.push_str(&orig_contents[..idx_start_code]); // comments block
	new_contents.push_str(MAGIC_PATCH);
	new_contents.push_str(&orig_contents[idx_start_code..]); // rest of file

	_write_replace_css_contents(&css_path, &new_contents)?;
	Ok(())
}

pub fn patch_icon(install_dir: &str) -> w::AnyResult<()> {
	let css_path = _build_css_path(install_dir);
	let orig_contents = _read_css_contents(&css_path)?;

	const NATURAL: &str = ".monaco-editor .suggest-widget .monaco-list .monaco-list-row.focused .codicon{color:var(--vscode-editorSuggestWidget-selectedIconForeground)}";
	const PATCHED: &str = " /*.monaco-editor .suggest-widget .monaco-list .monaco-list-row.focused .codicon{color:var(--vscode-editorSuggestWidget-selectedIconForeground)}*/ ";

	if let Some(_) = orig_contents.find(PATCHED) {
		return Err("Suggestion box icon already patched.\n\nNothing to do.".into());
	}

	let idx_part = match orig_contents.find(NATURAL) {
		Some(idx) => idx,
		None => return Err("Suggestion box icon CSS entry not found.".into()),
	};

	let mut new_contents = String::with_capacity(orig_contents.len() + PATCHED.len() - NATURAL.len());
	new_contents.push_str(&orig_contents[..idx_part]); // all code up to part
	new_contents.push_str(PATCHED);
	new_contents.push_str(&orig_contents[idx_part + NATURAL.len()..]); // rest of file

	_write_replace_css_contents(&css_path, &new_contents)?;
	Ok(())
}

fn _build_css_path(install_dir: &str) -> String {
	format!("{}\\{}",
		w::path::rtrim_backslash(install_dir),
		"resources\\app\\out\\vs\\workbench\\workbench.desktop.main.css")
}

fn _read_css_contents(css_path: &str) -> w::SysResult<String> {
	let fin = w::FileMapped::open(css_path, w::FileAccess::ExistingReadOnly)?;
	let s = w::WString::parse_str(fin.as_slice())?.to_string();
	Ok(s)
}

fn _write_replace_css_contents(css_path: &str, new_contents: &str) -> w::SysResult<()> {
	let fout = w::File::open(css_path, w::FileAccess::ExistingRW)?;
	fout.erase_and_write(new_contents.as_bytes())?;
	Ok(())
}

File: dlgs/patch.cpp
#include <optional>
#include "patch.h"
#include <CommCtrl.h>
#include <TlHelp32.h>

bool patch::is_vscode_running() {
	struct HandleSnap32 final {
		HANDLE h32;
		~HandleSnap32() noexcept { CloseHandle(h32); }
	};

	HandleSnap32 h{CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0)};
	if (h.h32 == INVALID_HANDLE_VALUE) [[unlikely]] {
		throw wd::WinErr{GetLastError(), L"CreateToolhelp32Snapshot failed."};
	}

	PROCESSENTRY32W pe{.dwSize = sizeof(PROCESSENTRY32W)};
	if (!Process32FirstW(h.h32, &pe)) {
		DWORD err = GetLastError();
		if (err == ERROR_NO_MORE_FILES) [[likely]] {
			return false;
		} else [[unlikely]] {
			throw wd::WinErr{err, L"Process32First failed."};
		}
	}

	for (;;) {
		if (wd::StrView{pe.szExeFile}.eq_i(L"Code.exe")) // process is running
			return true;

		if (!Process32NextW(h.h32, &pe)) {
			const DWORD err = GetLastError();
			if (err == ERROR_NO_MORE_FILES) [[likely]] {
				return false;
			} else [[unlikely]] {
				throw wd::WinErr{err, L"Process32Next failed."};
			}
		}
	}
}

static std::wstring patch_font(wd::StrView cssContents) {
	wd::StrView END_OF_COMMS = L"-*/";
	wd::StrView MAGIC_PATCH = L"\n*{text-shadow:transparent 0px 0px 0px, rgba(0, 0, 0, 0.5) 0px 0px 0px !important;}";

	size_t idxStartCode = cssContents.find(END_OF_COMMS);
	if (idxStartCode == wd::StrView::npos) [[unlikely]] {
		throw wd::WinErr{E_UNEXPECTED, L"CSS end of comments not found."};
	}
	idxStartCode += END_OF_COMMS.length();

	// Is our magic path the first thing past the comments block?
	if (cssContents.find(MAGIC_PATCH, idxStartCode) == idxStartCode) {
		throw wd::WinErr{E_INVALIDARG, L"Font already patched."};
	}

	std::wstring newContents{};
	newContents.reserve(cssContents.length() + MAGIC_PATCH.length());
	newContents.append(cssContents.begin(), cssContents.begin() + idxStartCode); // comments block
	newContents.append(MAGIC_PATCH);
	newContents.append(cssContents.begin() + idxStartCode, cssContents.end()); // rest of file

	return newContents;
}

static std::wstring patch_icon(wd::StrView cssContents) {
	wd::StrView NATURAL = L".monaco-editor .suggest-widget .monaco-list .monaco-list-row.focused .codicon{color:var(--vscode-editorSuggestWidget-selectedIconForeground)}";
	wd::StrView PATCHED = L" /*.monaco-editor .suggest-widget .monaco-list .monaco-list-row.focused .codicon{color:var(--vscode-editorSuggestWidget-selectedIconForeground)}*/ ";

	if (cssContents.contains(PATCHED)) {
		throw wd::WinErr{E_INVALIDARG, L"Suggestion box icon already patched."};
	}

	const size_t idx = cssContents.find(NATURAL);
	if (idx == wd::StrView::npos) [[unlikely]] {
		throw wd::WinErr{E_INVALIDARG, L"Suggestion box icon CSS entry not found."};
	}

	std::wstring newContents{};
	newContents.reserve(cssContents.length() + PATCHED.length() - NATURAL.length());
	newContents.append(cssContents.begin(), cssContents.begin() + idx); // all code up to part
	newContents.append(PATCHED);
	newContents.append(cssContents.begin() + idx + NATURAL.length(), cssContents.end()); // rest of file

	return newContents;
}

patch::Res patch::do_patch(wd::StrView installPath, bool doFont, bool doIcon) {
	const std::wstring cssPath = installPath + L"\\resources\\app\\out\\vs\\workbench\\workbench.desktop.main.css";
	bool fontOk = true, iconOk = true;
	Res res{
		.fontMsg = L"Font unpatched.",
		.iconMsg = L"Suggestion icon unpatched.",
	};

	std::wstring cssContents = wd::str::parse(wd::file::read(cssPath));
	if (doFont) {
		try {
			cssContents = patch_font(cssContents);
			res.fontMsg = L"Font patched successfully.";
		} catch (const wd::WinErr &e) {
			fontOk = false;
			res.fontMsg = e.msg();
		}
	}
	if (doIcon) {
		try {
			cssContents = patch_icon(cssContents);
			res.iconMsg = L"Suggestion icon patched successfully.";
		} catch (const wd::WinErr &e) {
			iconOk = false;
			res.iconMsg = e.msg();
		}
	}

	if (!fontOk && !iconOk) {
		res.tdIcon = TD_ERROR_ICON;
	} else if (!fontOk || !iconOk) {
		res.tdIcon = TD_WARNING_ICON;
	} else {
		res.tdIcon = TD_INFORMATION_ICON;
	}

	wd::file::write(cssPath, wd::str::to_utf8_blob(cssContents));
	return res;
}
